packages/integration_test/lib/common.dart

```diff
@@ -121,3 +132,105 @@ class Failure {
+enum WebDriverActionTypes {
+  /// Acknowlegement for the previously sent message.
+  ack,
+
+  /// No further WebDriver Action is necessary.
+  noop,
+
+  /// Asking WebDriver to take a screenshot of the Web page.
+  screenshot,
+}
+
+/// Command for WebDriver to execute.
+///
+/// Only works on Web when tests are run via `flutter driver` command.
+///
+/// See: https://www.w3.org/TR/webdriver/
+class WebDriverAction {
+  /// Type of the [WebDriverAction].
+  ///
+  /// Currently the only action that triggers a WebDriver API command is
+  /// `screenshot`.
+  ///
+  /// There are also `ack` and `noop` actions defined to manage the handshake
+  /// during the communication.
+  final WebDriverActionTypes type;
+
+  /// Used for adding extra values to the actions such as file name for
+  /// `screenshot`.
+  final Map<String, dynamic> values;
+
+  /// Constructor for [WebDriverActionTypes.noop] action.
+  WebDriverAction.noop()
+      : this.type = WebDriverActionTypes.noop,
+        this.values = Map();
+
+  /// Constructor for [WebDriverActionTypes.noop] screenshot.
+  WebDriverAction.screenshot(String screenshot_name)
+      : this.type = WebDriverActionTypes.screenshot,
+        this.values = {'screenshot_name': screenshot_name};
+
+  /// Util method for converting [WebDriverActionTypes] to a map entry.
+  ///
+  /// Used for converting messages to json format.
+  static Map<String, dynamic> typeToMap(WebDriverActionTypes type) => {
+        'web_driver_action': '${type}',
+      };
+}
+
+/// Template methods to implement for any class which manages communication
+/// between `integration_tests` and the `driver_tests`.
+///
+/// See example [WebDriverCommandManager].
+abstract class DriverCommandManager {
+  /// The callback function to response the driver side input which can also
+  /// send WebDriver command requests such as `screenshot` to driver side.
+  Future<Map<String, dynamic>> callbackWithDriverCommands(
+      Map<String, String> params, IntegrationTestResults testRunner);
+
+  /// Request to take a screenshot of the application from the driver side.
+  void takeScreenshot(String screenshot);
+
+  /// Cleanup and completers or locks used during the communication.
+  void cleanup();
+}
+
+/// Interface that surfaces test results of integration tests.
+///
+/// Implemented by [IntegrationTestWidgetsFlutterBinding]s.
+///
+/// Any class which needs to access the test results but do not want to create
+/// a cyclic dependency [IntegrationTestWidgetsFlutterBinding]s can use this
+/// interface. Example [WebDriverCommandManager].
+abstract class IntegrationTestResults {
+  /// Stores failure details.
+  ///
+  /// Failed test method's names used as key.
+  List<Failure> get failureMethodsDetails;
+
+  /// The extra data for the reported result.
+  Map<String, dynamic> get reportData;
+
+  /// Whether all the test methods completed succesfully.
+  Completer<bool> get allTestsPassed;
+}
```

Comment:
I'm a bit apprehensive about exporting the methods here as part of the public API of IntegrationTestWidgetsFlutterBinding. There are some [plans](https://github.com/flutter/flutter/issues/50816) to change the test reporting API and it will be good to keep such changes internal, at least for now. What about changing this interface to something the following to limit the number of public methods we expose?

```
class ResponseProvider {
  Response get reponse;
}
```

 I think the Response object above has all the methods you need here

Reply:
Unless there is an alternative of avoiding cycling dependency between CommandManager's and the binding I keep using this. All the methods are getter so the reader won't be able to access the values of the test results.

I think this is more of a language limitation, where we are missing a package level modifier.